keps/sig-storage/2644-honor-pv-reclaim-policy/README.md

```diff
@@ -172,16 +178,39 @@ the current behavior.
+
+When CSI Migration is enabled, external-provisioner adds `external-provisioner.volume.kubernetes.io/finalizer` to all the PVs it creates, including in-tree ones. When CSI Migration is disabled, however, these PVs will be deleted by in-tree volume plugin. Therefore, in-tree PV controller needs to be modified to remove the finalizer when the PV is being deleted when CSI Migration is disabled.
 
 ```go
 // PVDeletionProtectionFinalizer finalizer is added to the PV to prevent PV from being deleted before the physical volume
 PVDeletionProtectionFinalizer = "external-provisioner.volume.kubernetes.io/finalizer"
 ```
 
-In `deleteVolumeOperation` in kubernetes/pkg/controller/volume/persistentvolume/pv_controller.go, PV without `PVDeletionProtectionFinalizer` will be skipped as it is already processed and PV with `PVDeletionProtectionFinalizer` will proceed with volume deletion. This will allow external-provisioner to handle the deletion of the CSI volumes. The finalizer `PVDeletionProtectionFinalizer` will only be deleted after the underlying CSI volume is deleted.
+On the CSI driver side, the library adds the finalizer only to newly provisioned PVs. The proposal is to extend the library to (optionally) add the finalizer to all PVs (that are handled by the external-provisioner), to have all PVs protected once the feature is enabled.
+
+When the `shouldDelete` checks succeed, a delete volume request is initiated on the driver. This ensures that the volume is deleted on the backend.
+
+Once the volume is deleted from the backend, the finalizer can be removed. This allows the pv to be removed from the api server.
+
+Note: This feature should work with CSI Migration disabled or enabled.
+
+### In-Tree Plugin volumes
+
+A new finalizer `kubernetes.io/pv-controller` will be introduced. The finalizer will be added to newly created in-tree volumes as well as existing in-tree volumes. The finalizer will only be removed after the plugin successfully deletes the in-tree volume.
+
+When CSI Migration is enabled, the finalizer `kubernetes.io/pv-controller` will be removed from the in-tree volume PV, and as stated previously, the external-provisioner adds `external-provisioner.volume.kubernetes.io/finalizer` finalizer on to the PV. However, when CSI Migration is disabled, the finalizer `kubernetes.io/pv-controller` is added back on the PV.
+
+```go
+    //PVDeletionInTreeProtectionFinalizer is the finalizer added to protect PV deletion for in-tree volumes.
+	PVDeletionInTreeProtectionFinalizer = "kubernetes.io/pv-controller"
+
+    // Please refer pkg/controller/volume/persistentvolume/util/util.go for the full code.
+```
+
+In `deleteVolumeOperation` in kubernetes/pkg/controller/volume/persistentvolume/pv_controller.go, checks for `DeletionTimestamp` should be ignored, the check currently prevents the volume from being deleted by the underlying storage by the plugin. In cases where the PV is deleted first, the `DeletionTimestamp` is expected to be set on the PV.
```

Comment:
It may happen that a PV is without any finalizer when in-tree PV controller removed its finalizer and the external-provisioner did not add its own to the PV. Such a volume may be deleted by the API server.

I don't think it's an issue, still, it would be nice to mention it explicitly.

Reply:
Done.